akka-http-core/src/main/scala/akka/http/impl/engine/parsing/BodyPartParser.scala

```diff
@@ -57,202 +60,216 @@ private[http] final class BodyPartParser(defaultContentType: ContentType,
-          // we are missing a try/catch{continue} wrapper somewhere
-          throw new IllegalStateException("unexpected NotEnoughDataException", NotEnoughDataException)
+  val in = Inlet[ByteString]("BodyPartParser.in")
+  val out = Outlet[BodyPartParser.Output]("BodyPartParser.out")
+
+  override val shape = FlowShape(in, out)
+
+  override def createLogic(attributes: Attributes): GraphStageLogic =
+    new GraphStageLogic(shape) with InHandler with OutHandler {
+      private var output = collection.immutable.Queue.empty[Output]
+      // FIXME this probably is too wasteful
+      private var state: ByteString => StateResult = tryParseInitialBoundary
+      private var terminated = false
+
+      override def onPush(): Unit = {
+        if (!terminated) {
+          val elem = grab(in)
+          try state(elem)
+          catch {
+            case e: ParsingException => fail(e.info)
+            case NotEnoughDataException =>
+              // we are missing a try/catch{continue} wrapper somewhere
+              throw new IllegalStateException("unexpected NotEnoughDataException", NotEnoughDataException)
+          }
+          if (output.nonEmpty) push(out, dequeue())
+          else if (!terminated) pull(in)
+          else completeStage()
+        } else completeStage()
       }
-      if (output.nonEmpty) ctx.push(dequeue())
-      else if (!terminated) ctx.pull()
-      else ctx.finish()
-    } else ctx.finish()
-
-  override def onPull(ctx: Context[Output]): SyncDirective = {
-    if (output.nonEmpty)
-      ctx.push(dequeue())
-    else if (ctx.isFinishing) {
-      if (terminated) ctx.finish()
-      else ctx.pushAndFinish(ParseError(ErrorInfo("Unexpected end of multipart entity")))
-    } else
-      ctx.pull()
-  }
-
-  override def onUpstreamFinish(ctx: Context[Output]): TerminationDirective = ctx.absorbTermination()
-
-  def tryParseInitialBoundary(input: ByteString): StateResult =
-    // we don't use boyerMoore here because we are testing for the boundary *without* a
-    // preceding CRLF and at a known location (the very beginning of the entity)
-    try {
-      if (boundary(input, 0)) {
-        val ix = boundaryLength
-        if (crlf(input, ix)) parseHeaderLines(input, ix + 2)
-        else if (doubleDash(input, ix)) terminate()
-        else parsePreamble(input, 0)
-      } else parsePreamble(input, 0)
-    } catch {
-      case NotEnoughDataException ⇒ continue(input, 0)((newInput, _) ⇒ tryParseInitialBoundary(newInput))
-    }
 
-  def parsePreamble(input: ByteString, offset: Int): StateResult =
-    try {
-      @tailrec def rec(index: Int): StateResult = {
-        val needleEnd = boyerMoore.nextIndex(input, index) + needle.length
-        if (crlf(input, needleEnd)) parseHeaderLines(input, needleEnd + 2)
-        else if (doubleDash(input, needleEnd)) terminate()
-        else rec(needleEnd)
+      override def onPull(): Unit = {
+        if (output.nonEmpty)
+          push(out, dequeue())
```

Comment:
This formatting seems a bit off with the linefeed after if like this, did scalariform do it? (if not remove linefeed!)